protocols/rendezvous/src/client.rs

```diff
@@ -218,159 +225,187 @@ impl NetworkBehaviour for Behaviour {
         connection_id: ConnectionId,
         event: THandlerOutEvent<Self>,
     ) {
-        let new_events = match event {
-            handler::OutboundOutEvent::InboundEvent { message, .. } => void::unreachable(message),
-            handler::OutboundOutEvent::OutboundEvent { message, .. } => handle_outbound_event(
-                message,
-                peer_id,
-                &mut self.discovered_peers,
-                &mut self.expiring_registrations,
-            ),
-            handler::OutboundOutEvent::InboundError { error, .. } => void::unreachable(error),
-            handler::OutboundOutEvent::OutboundError { error, .. } => {
-                log::warn!("Connection with peer {} failed: {}", peer_id, error);
-
-                vec![ToSwarm::CloseConnection {
-                    peer_id,
-                    connection: CloseConnection::One(connection_id),
-                }]
-            }
-        };
+        self.inner
+            .on_connection_handler_event(peer_id, connection_id, event);
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        self.external_addresses.on_swarm_event(&event);
 
-        self.events.extend(new_events);
+        self.inner.on_swarm_event(event);
     }
 
     fn poll(
         &mut self,
         cx: &mut Context<'_>,
-        _: &mut impl PollParameters,
+        params: &mut impl PollParameters,
     ) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
-        if let Some(event) = self.events.pop_front() {
-            return Poll::Ready(event);
+        if let Some(event) = self.error_events.pop_front() {
+            return Poll::Ready(ToSwarm::GenerateEvent(event));
         }
 
-        if let Some((namespace, rendezvous_node, ttl)) = self.pending_register_requests.pop() {
-            // Update our external addresses based on the Swarm's current knowledge.
-            // It doesn't make sense to register addresses on which we are not reachable, hence this should not be configurable from the outside.
-
-            let external_addresses = self.external_addresses.iter().cloned().collect::<Vec<_>>();
+        let poll_res = self.inner.poll(cx, params);
 
-            if external_addresses.is_empty() {
-                return Poll::Ready(ToSwarm::GenerateEvent(Event::RegisterFailed(
-                    RegisterError::NoExternalAddresses,
-                )));
+        if let Poll::Ready(ToSwarm::GenerateEvent(event)) = &poll_res {
+            if let libp2p_request_response::Event::Message {
+                peer: _,
+                message: libp2p_request_response::Message::Request { .. },
+            } = event
+            {
+                return Poll::Pending;
```

Comment:
This breaks the poll-fn contract, we should loop instead until our inner behaviour returns pending.